staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,1011 @@
+// feature gate.
+type ResourceClaim struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec describes what is being requested and how to configure it.
+	// The spec is immutable.
+	Spec ResourceClaimSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+
+	// Status describes whether the claim is ready to use and what has been allocated.
+	// +optional
+	Status ResourceClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ResourceClaimSpec defines what is being requested in a ResourceClaim and how to configure it.
+type ResourceClaimSpec struct {
+	// Devices defines how to request devices.
+	//
+	// +optional
+	Devices DeviceClaim `json:"devices" protobuf:"bytes,1,name=devices"`
+
+	// Controller is the name of the DRA driver that is meant
+	// to handle allocation of this claim. If empty, allocation is handled
+	// by the scheduler while scheduling a pod.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	Controller string `json:"controller,omitempty" protobuf:"bytes,2,opt,name=controller"`
+}
+
+// DeviceClaim defines how to request devices with a ResourceClaim.
+type DeviceClaim struct {
+	// Requests represent individual requests for distinct devices which
+	// must all be satisfied. If empty, nothing needs to be allocated.
+	//
+	// +optional
+	// +listType=atomic
+	Requests []DeviceRequest `json:"requests" protobuf:"bytes,1,name=requests"`
+
+	// These constraints must be satisfied by the set of devices that get
+	// allocated for the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Constraints []DeviceConstraint `json:"constraints,omitempty" protobuf:"bytes,2,opt,name=constraints"`
+
+	// This field holds configuration for multiple potential drivers which
+	// could satisfy requests in this claim. It is ignored while allocating
+	// the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []DeviceClaimConfiguration `json:"config,omitempty" protobuf:"bytes,3,opt,name=config"`
+}
+
+// DeviceRequest is a request for devices required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices.
+//
+// A DeviceClassName is currently required. Clients must check that it is
+// indeed set. It's absence indicates that something changes in a way that
+// is not supported by the client yet, in which case it must refuse to
+// handle the request.
+type DeviceRequest struct {
+	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
+	// entry and in a constraint of the claim.
+	//
+	// Must be a DNS label.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
```

Comment:
https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1659413405 has become unlinked (yay github!) but is one of the more important threads in this PR.

The topic at hand was how "one of the following device shapes" (as opposed to "this specific device shape") will fit.

As it is currently defined, the embedded `DeviceRequestDetails.DeviceClassName` is marked required.  If we leave it required, then we can't really have a on-of alternative to it.  If we make it optional, then it's not required (as we want).  If we make it a one-of, then reusing this type doesn't work.  It's  starting to feel like "have your cake and eat it, too".

I think we will end up with extra nesting or 2 definitions of the same schema.

Urban planning style:
```
type DeviceRequest struct {
        // +required
        Name string 

        // +optional
        // +oneOf=RequestMode
        Specific *SpecificDeviceRequest

        // optional
        // +oneOf=RequestMode
        OneOf []SpecificDeviceRequest
}

type SpecificDeviceRequest struct {                                                                                                                                               
        // +required
        DeviceClassName string `json:"deviceClassName" protobuf:"bytes,1,name=deviceClassName"`

        // +optional
        Selectors []DeviceSelector `json:"selectors,omitempty" protobuf:"bytes,2,name=selectors"`

        // +optional
        CountMode DeviceCountMode `json:"countMode,omitempty" protobuf:"bytes,3,opt,name=countMode"`

        // +optional
        Count int64 `json:"count,omitempty" protobuf:"bytes,4,opt,name=count"`

        // +optional
        // +default=false
        AdminAccess bool `json:"adminAccess,omitempty" protobuf:"bytes,5,opt,name=adminAccess"`
}
```

...or else...

Organic evolution style:
```
type DeviceRequest struct {
        // +required
        Name string 

        // +optional
        // +oneOf=RequestMode
        DeviceClassName string 

        // +optional
        // Only valid if DeviceClassName is specified.
        Selectors []DeviceSelector 

        // +optional
        // Only valid if DeviceClassName is specified.
        CountMode DeviceCountMode 

        // +optional
        // Only valid if DeviceClassName is specified.
        Count int64 

        // +optional
        // +default=false
        // Only valid if DeviceClassName is specified.
        AdminAccess bool 

        // optional
        // +oneOf=RequestMode
        OneOf []SpecificDeviceRequest
}

type SpecificDeviceRequest struct {                                                                                                                                               
        // +required
        DeviceClassName string `json:"deviceClassName" protobuf:"bytes,1,name=deviceClassName"`

        // +optional
        Selectors []DeviceSelector `json:"selectors,omitempty" protobuf:"bytes,2,name=selectors"`

        // +optional
        CountMode DeviceCountMode `json:"countMode,omitempty" protobuf:"bytes,3,opt,name=countMode"`

        // +optional
        Count int64 `json:"count,omitempty" protobuf:"bytes,4,opt,name=count"`

        // +optional
        // +default=false
        AdminAccess bool `json:"adminAccess,omitempty" protobuf:"bytes,5,opt,name=adminAccess"`
}
```

Reply:
I agree with the analysis. So which way do we want to go? :thinking: 

I think I am leaning towards the "Organic evolution style" because the resulting API for direct requests (the common case) is simpler.

Reply:
@klueska: your preference?

Reply:
I spoke with @klueska earlier today and he agreed with using the "organic evolution style".

I went ahead and made the changes. It simplified Go code quite a bit (initializing embedded structs means repeating the struct name), which reassures me that we are one the right track.

For now all my changes are in incremental commits. I'll squash later.

Reply:
Reopened because of https://github.com/kubernetes/kubernetes/pull/125953#discussion_r1669188760, we can close again if we agree to go ahead as implemented right now despite that.

Suppose we want to add a new field alongside `DeviceClassName`, like a request to consume some consumable resources of a device. We can't just add such a field because old clients would ignore it.

With the current API, we would need to define a "ConsumeableDeviceRequest", add all current fields and the new one there, then add a `Consume *ConsumeableDeviceRequest` field as alternative to the ones we have right now.

With that in mind, should we add a `RequestMode` discriminator after all? Honest question, I don't know.

/cc @johnbelamaric

Reply:
If we add a `RequestMode` enum, we still need a name for the current mode. `Simple`? How would it be explained what that means? Right now it has to be "as defined in 1.31". Later, it could be extended to say "without using field(s) XYZ".

Reply:
The more I think about it, the less I like the discriminator approach. A separate struct will be much more explicit.

Reply:
> With the current API, we would need to define a "ConsumeableDeviceRequest", add all current fields and the new one there, then add a Consume *ConsumeableDeviceRequest field as alternative to the ones we have right now.

That sounds pretty awful. If we add yet-another-field, we get yet-another-struct. This is exactly the paired problem we just (hopefully) resolved on the capacity problem. There, we used a top level one-of; here we have a sort of one-of-with-inlined-option. I think either of those is bad here, though, because this is user-facing and it would be awkward for users to have to pick their one-of name (or a discriminator) as they decide which fields they want to use in their request. Hmm...

So, [there are few things we know](https://docs.google.com/document/d/1tvdBniYMWo79hrIeCYkJsUylzvh90xSOC9dm-rLAk88/edit?pli=1) that we *might* want to add to the requests side. In "order of likelihood" (as I see it):
- Prioritized list of ways to satisfy the request ("give me this or that or 2 of those")
- Ranged count with aggregate resource criteria ("give me 1-4 with aggregate memory of 128Gi")
- Per-device allocatable resources ("give me 10Gi and 1 core from any device")
- Access mode (shared read, write with shared read, exclusive)
- Linked claims (allocate a set of devices in one claim, then "attach" another claim to it; either admin mode or in a shared access mode) 

How would we support each?
- prioritized list is handled as described above, with the one-of-with-inlined-option.
- ranged count with aggregate resource criteria would be handled with:
  - `CountMode = Range`
  - A new constraint type, like `AggregateCapacity` or something (`claim.devices[*].constraints[*].aggregateCapacity['memory'] = 128Gi`)
- Per-device allocatable resources - could we do this with something like `CountMode = Fractional` along with a new `Requests map[string]Quantity` field, to explicitly allow fractional devices? If you want "whole device" you can use capacity in your CEL expression; you only need `requests` if you want fractional devices.
- Access mode could probably be another type of constraint, so it fits in that one-of.
- Linked claims could be an alternative in the DeviceSelector one-of.

So, I guess what I am saying is, we have a Selector one-of, and we have a CountMode discriminator, and we have a Constraints one-of. That may actually be enough to capture everything we can think of right now, on the claim side.

It does make me nervous though that we could box ourselves in without *any* sort of discriminator field now. What if we add one that has a default and only one valid value, just to be safe?

Reply:
@johnbelamaric if I understand correctly, you are proposing we go ahead and add these fields *now*, but don't actually support them yet (e.g. by making validation error out if they are set).

Reply:
> It does make me nervous though that we could box ourselves in without any sort of discriminator field now. What if we add one that has a default and only one valid value, just to be safe?

I'm not convinced. @johnbelamaric, presumably you then want to bump that string to something else when adding a new field in the `Request struct` instead of going the "new struct alternative" way. Correct?

That is exactly the approach that was rejected by @thockin for ResourceSlice because a) [it is not done anywhere else](https://github.com/kubernetes/kubernetes/pull/125953#discussion_r1669063885) and b) it makes it hard to reason about which fields are valid and/or have which semantic (not sure where it was said, but he mentioned other APIs where some fields may only be set under certain conditions).

Reply:
> you are proposing we go ahead and add these fields now, but don't actually support them yet (e.g. by making validation error out if they are set).

I think he just wants a `RequestType string` field with a single valid value. If a client sees a different value, they know that they don't support the request.

Reply:
> @johnbelamaric if I understand correctly, you are proposing we go ahead and add these fields now, but don't actually support them yet (e.g. by making validation error out if they are set).

No, I was trying to game out "can we make the things we know about work with the API we have now", and I *think* the answer I came up with is "yes".

> I think he just wants a RequestType string field with a single valid value. If a client sees a different value, they know that they don't support the request.

Right - just for "the things we do not yet know about".

Reply:
> I'm not convinced. @johnbelamaric, presumably you then want to bump that string to something else when adding a new field in the Request struct instead of going the "new struct alternative" way. Correct?

To be clear, I think the "new struct" makes sense for the "one-of" functionality, and that we have other levers and means for the evolutions listed above (as described above). The "RequestType" is there *just in case* we need to do something that doesn't make sense via one of the other means that we have discussed; it's the final escape hatch.

Reply:
I called this "discriminator" earlier, but I think that wasn't quite the right term: a "descriminator" has a clearly defined scope and probably already more than one valid value from the start, like our `CountMode`.

What we thought about here is more like a "dead man switch" or "sentinel". It doesn't really have a meaning yet other than "must not change" - that's the new thing that isn't done anywhere else. It's also not clear which struct needs it and which doesn't, unless the API is clear from the beginning.

@johnbelamaric: are you okay with dropping that idea and not adding such a field? I think then we have consensus and can close this thread again.

Reply:
Yes, I am OK without that extra field.

Reply:
I agree that we are okay. 

The reason why I am trying to make it possible for clients to detect that they are out-dated is because I expect that there will be several out-of-tree components which need to work with requests. The developers of those components might not follow future DRA extensions closely enough to update their components. I think it's better if such an out-dated component fails clearly or at least logs something instead of silently not working right, even if "not right" isn't too bad.

Having said that, we should only do that if it doesn't make the current API too cumbersome. I think we have found a good balance.

> We might want to rename "CountMode" to "AllocationMode" to be clear that within this specific device request, that is the discriminator.

Agreed. I'll make that change. Note that I am changing "Exact" to "ExactCount": `allocationMode: ExactCount`.

Reply:
See https://github.com/kubernetes/kubernetes/compare/a537356172a4b1f5d87fa215a4620dd0e471e5fe..bda7cfc311fe3d561612676ee65e170888e31b82 for a diff.

I'm following up with a fix for some failing tests (not all of them got updated properly when making some changes), but that won't have other changes worth reviewing.

Reply:
ACK